### PR TITLE
Use Foq's return strategy in AutoFoq

### PR DESCRIPTION
**Pull request by moodmosaic**
Related to the issue described in #214.

### CODE REVIEW

File: Src/AutoFoqUnitTest/FoqMethodQueryTest.fs [hunk 1]
```diff
@@ -87,6 +105,17 @@ let SelectMethodsDefineCorrectParameters (request: Type) =
...
+[<Fact>]
+let ``Builder is correct`` () =
+    let expected =
+        { new ISpecimenBuilder with
+              member this.Create(r, c) = obj() }
+    let sut = FoqMethodQuery expected
+
+    let actual = sut.Builder
+
+    verify <@ expected = actual @>
```
Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FoqMethodQueryTest.fs]: I've just noticed, by looking at the Pull Request, that this can be changed to `expected =? actual`.

Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FoqMethodQueryTest.fs]: Actually, I've just noticed that we're using an old version of Unquote, and these old operators are deprecated in newer versions, [as shown in the Unquote source code](http://git.io/v8KPF):

```
///These '?' suffixed operators conflict with F# 3.0's nullable operators and
/// have been replaced by equivalent '!' suffixed operators.

[<System.Obsolete(
    "These '?' suffixed operators conflict with F# 3.0's nullable operators and
     have been replaced by equivalent '!' suffixed operators.")>]
module Obsolete =
    let inline (=?) x y = test <@ x = y @>
    // ...

let inline (=!) x y = test <@ x = y @>
//...
```

So I'll leave this one as-is, but I'll change an other test added in this PR so it doesn't use that deprecated `=?` operator.

**Tl;dr**

We should consider upgrading Unquote, in a separate Pull Request.

Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FoqMethodQueryTest.fs]: Added 3a5e842, in order to address [this](https://github.com/AutoFixture/AutoFixture/pull/464#discussion_r44374058) comment of mine. – Done adding commits.

File: Src/AutoFoq/AutoFoqCustomization.fs [hunk 2]
```diff
@@ -18,7 +18,8 @@ type AutoFoqCustomization(relay: ISpecimenBuilder) =
         AutoFoqCustomization(
             FilteringSpecimenBuilder(
                 MethodInvoker(
-                    FoqMethodQuery()),
+                    FoqMethodQuery(
+                        Fixture())),
```
Comment (ploeh) [inline Src/AutoFoq/AutoFoqCustomization.fs]: An `ICustomization` is normally something you apply to a given `IFixture` instance. Wouldn't this mean that the default `AutoFoqCustomization` applied to an `IFixture` would contain *another* `Fixture` instance? Isn't that going to be confusing to users?

File: Src/AutoFoqUnitTest/FixtureIntegrationTest.fs [hunk 3]
```diff
@@ -66,3 +67,29 @@ let FixtureSuppliesValuesToAbstractGenericTypeWithNonDefaultConstructorWithMulti
...
+    let sut = (Fixture().Customize(AutoFoqCustomization())).Create<IInterface>()
+    let actual = sut.ReturnSomething()
+    verify <@ not (isNull actual) @>
+
+[<Fact>]
+let ``Fixture customizations are propagated in Foq for non-explicitly setup test-doubles`` () =
+    let fixture = Fixture()
+    let expected = "Friday, 14 December 1984"
+    fixture.Register<string>(fun () -> expected)
+    fixture.ResidueCollectors.Add(
```
Comment (ploeh) [inline Src/AutoFoqUnitTest/FixtureIntegrationTest.fs]: Shouldn't this instead add `AutoFoqCustomization` in order to demonstrate that other Customizations take effect through Foq?

File: Src/AutoFoqUnitTest/FoqMethodQueryTest.fs [hunk 4]
```diff
@@ -74,7 +89,10 @@ let SelectMethodsDefineCorrectParameters (request: Type) =
             ||| BindingFlags.Instance
             ||| BindingFlags.NonPublic)         
         |> Seq.map(fun ci -> ci.GetParameters())
-    let sut = FoqMethodQuery()
+    let dummyBuilder =
```
Comment (ploeh) [inline Src/AutoFoqUnitTest/FoqMethodQueryTest.fs]: Would it make sense to make `dummyBuilder` a module value?

File: Src/AutoFoq/AutoFoqCustomization.fs (hunk 2 above)
Comment (moodmosaic) [inline Src/AutoFoq/AutoFoqCustomization.fs]: To be honest, I was going a bit back and forth on that one :neutral_face:

In the beginning I tried to move the whole object graph in the `Customize` method, in order to supply the given `IFixture` instance, but then I got failing tests which were not easy to make them passing.

---

Then I thought that supplying a *default* `Fixture` can be considered an implementation detail; just so that Foq has something reasonable to return. – After all, this affects *only* those Test Doubles that have not being explicitly setup.

So I went by supplying a default `Fixture` instance, but I've also added a façade test [here](https://github.com/moodmosaic/AutoFixture/blob/use-autofoq-return-strategy-to-avoid-nulls/Src/AutoFoqUnitTest/FixtureIntegrationTest.fs#L82-L95) to demonstrate that it's possible to use the types defined in AutoFoq with a customized `Fixture` instance.

File: Src/AutoFoqUnitTest/FixtureIntegrationTest.fs (hunk 3 above)
Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FixtureIntegrationTest.fs]: This happens already in that other test, above this one. This one exists mostly for the reason described [here](https://github.com/AutoFixture/AutoFixture/pull/464/files#r44875292).

File: Src/AutoFoqUnitTest/FoqMethodQueryTest.fs (hunk 4 above)
Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FoqMethodQueryTest.fs]: Yes, I think it would make sense :+1:

File: Src/AutoFoqUnitTest/FixtureIntegrationTest.fs (hunk 3 above)
Comment (ploeh) [inline Src/AutoFoqUnitTest/FixtureIntegrationTest.fs]: > This happens already in that other test, above this one.

They aren't testing the same behaviour. The first one only checks to see if `actual` isn't null. This one is more explicit.

File: Src/AutoFoq/AutoFoqCustomization.fs (hunk 2 above)
Comment (ploeh) [inline Src/AutoFoq/AutoFoqCustomization.fs]: I think we need to find a solution to this. Having *two* Fixture instances in play is bound to blow up in our faces sooner or later.

Comment (moodmosaic) [inline Src/AutoFoq/AutoFoqCustomization.fs]: Agreed, let's do it. After all, since this PR targets the `v4` branch things are a bit more flexible.

File: Src/AutoFoqUnitTest/FixtureIntegrationTest.fs (hunk 3 above)
Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FixtureIntegrationTest.fs]: [I'll get rid of that default `Fixture` instance](https://github.com/AutoFixture/AutoFixture/pull/464/files#r44896077) in order to make it possible to just add the `AutoFoqCustomization`.

File: Src/AutoFoqUnitTest/FoqMethodQueryTest.fs (hunk 4 above)
Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FoqMethodQueryTest.fs]: Done in [ef5280e](https://github.com/moodmosaic/AutoFixture/commit/ef5280ed4a30827caaf1ee33999937b7b9c7781a).

File: Src/AutoFoq/AutoFoqCustomization.fs (hunk 2 above)
Comment (moodmosaic) [inline Src/AutoFoq/AutoFoqCustomization.fs]: Done in [3702089](https://github.com/AutoFixture/AutoFixture/commit/3702089b1bd10cfd5c2d7221949636a01e367cc0).

**Comment by moodmosaic** [comment]
Review comments addressed. – Done adding commits.

File: Src/AutoFoqUnitTest/FixtureIntegrationTest.fs [hunk 5]
```diff
@@ -66,3 +67,24 @@ let FixtureSuppliesValuesToAbstractGenericTypeWithNonDefaultConstructorWithMulti
...
+let ``Fixture customizations are propagated in Foq for non-explicitly setup test-doubles`` () =
+    let fixture = Fixture()
+    let expected = "Friday, 14 December 1984"
+    fixture.Register<string>(fun () -> expected)
+    fixture.Customize(AutoFoqCustomization()) |> ignore
+    let sut = fixture.Create<IInterface>()
+
+    let actual = sut.ReturnSomething()
+
+    verify <@ expected = actual @>
```
Comment (ploeh) [inline Src/AutoFoqUnitTest/FixtureIntegrationTest.fs]: Fewer unit tests (deletions from `AutoFoqCustomizationTest`) and more [Facade Tests](http://blog.ploeh.dk/2012/06/27/FacadeTest): yay :+1:

**Comment by ploeh** [comment]
Thank you. It's merged into *v4* only, since it contains breaking changes. AFAIU, that was also your intent all along.

File: Src/AutoFoqUnitTest/FixtureIntegrationTest.fs (hunk 5 above)
Comment (moodmosaic) [inline Src/AutoFoqUnitTest/FixtureIntegrationTest.fs]: :wink:

**Comment by moodmosaic** [comment]
Thank you for a great review! Yes, that's correct, this one is for the v4 branch.
